params/mutations/rewards_test.go

```diff
@@ -192,3 +214,335 @@ func TestGetBlockWinnerRewardForUnclesByEra(t *testing.T) {
+		if err != nil {
+			t.Fatalf("could not open statedb: %v", err)
+		}
+
+		var header *types.Header = &types.Header{}
+		var uncles []*types.Header = []*types.Header{{}, {}}
+
+		if i == 0 {
+			header.Coinbase = common.HexToAddress("000d836201318ec6899a67540690382780743280")
+			uncles[0].Coinbase = common.HexToAddress("001762430ea9c3a26e5749afdb70da5f78ddbb8c")
+			uncles[1].Coinbase = common.HexToAddress("001d14804b399c6ef80e64576f657660804fec0b")
+		} else {
+			header.Coinbase = common.HexToAddress("0000000000000000000000000000000000000001")
+			uncles[0].Coinbase = common.HexToAddress("0000000000000000000000000000000000000002")
+			uncles[1].Coinbase = common.HexToAddress("0000000000000000000000000000000000000003")
+		}
+
+		// Manual tallies for reward accumulation.
+		winnerB, totalB := new(big.Int), new(big.Int)
+		unclesB := []*big.Int{new(big.Int), new(big.Int)}
+
+		winnerB = stateDB.GetBalance(header.Coinbase)
+		unclesB[0] = stateDB.GetBalance(uncles[0].Coinbase)
+		unclesB[1] = stateDB.GetBalance(uncles[1].Coinbase)
+
+		totalB.Add(totalB, winnerB)
+		totalB.Add(totalB, unclesB[0])
+		totalB.Add(totalB, unclesB[1])
+
+		if totalB.Cmp(big.NewInt(0)) != 0 {
+			t.Errorf("unexpected: %v", totalB)
+		}
+
+		for _, c := range cases[i] {
+			bn := c.block
+			era := GetBlockEra(bn, eraLen)
+
+			header.Number = bn
+
+			for i, uncle := range uncles {
+
+				// Randomize uncle numbers with bound ( n-1 <= uncleNum <= n-7 ), where n is current head number
+				// See yellowpaper@11.1 for ommer validation reference. I expect n-7 is 6th-generation ommer.
+				// Note that ommer nth-generation impacts reward only for "Era 1".
+				rand.Seed(time.Now().UTC().UnixNano())
+
+				// 1 + [0..rand..7) == 1 + 0, 1 + 1, ... 1 + 6
+				un := new(big.Int).Add(big.NewInt(1), big.NewInt(int64(rand.Int31n(int32(7)))))
+				uncle.Number = new(big.Int).Sub(header.Number, un) // n - un
+
+				ur := GetBlockUncleRewardByEra(era, header, uncle, bn)
+				unclesB[i].Add(unclesB[i], ur)
+
+				totalB.Add(totalB, ur)
+			}
+
+			wr := GetBlockWinnerRewardByEra(era, bn)
+			wr.Add(wr, GetBlockWinnerRewardForUnclesByEra(era, uncles, bn))
+			winnerB.Add(winnerB, wr)
+
+			totalB.Add(totalB, winnerB)
+
+			AccumulateRewards(config, stateDB, header, uncles)
+
+			// Check balances.
+			//t.Logf("config=%d block=%d era=%d w:%d u1:%d u2:%d", i, bn, new(big.Int).Add(era, big.NewInt(1)), winnerB, unclesB[0], unclesB[1])
+			if wb := stateDB.GetBalance(header.Coinbase); wb.Cmp(winnerB) != 0 {
+				t.Errorf("winner balance @ %v, want: %v, got: %v (config: %v)", bn, winnerB, wb, i)
+			}
+			if uB0 := stateDB.GetBalance(uncles[0].Coinbase); unclesB[0].Cmp(uB0) != 0 {
+				t.Errorf("uncle1 balance @ %v, want: %v, got: %v (config: %v)", bn, unclesB[0], uB0, i)
+			}
+			if uB1 := stateDB.GetBalance(uncles[1].Coinbase); unclesB[1].Cmp(uB1) != 0 {
+				t.Errorf("uncle2 balance @ %v, want: %v, got: %v (config: %v)", bn, unclesB[1], uB1, i)
+			}
+		}
+		db.Close()
+	}
+}
+*/
```

Comment:
This test (from the old code), still needs resolving. It runs (after refactor), but fails massively, further study required.

Reply:
resolved.